File: tests/fixtures/real_world/javascript/safe/safe_parameterized_variants.js
const mysql = require('mysql');
const { Client, Pool } = require('pg');

// MySQL ? placeholder with connection.query
function mysqlSafe(req, res) {
    const id = req.query.id;
    const connection = mysql.createConnection({host: 'localhost'});
    connection.query('SELECT * FROM users WHERE id = ?', [id], function(err, rows) {
        res.json(rows);
    });
}

// PostgreSQL $1/$2 placeholders with client.query
async function pgClientSafe(req, res) {
    const name = req.query.name;
    const age = req.query.age;
    const client = new Client();
    const result = await client.query('SELECT * FROM users WHERE name = $1 AND age = $2', [name, age]);
    res.json(result.rows);
}

// Pool.query with $1 placeholder
async function pgPoolSafe(req, res) {
    const id = req.params.id;
    const pool = new Pool();
    const result = await pool.query('SELECT email FROM users WHERE id = $1', [id]);
    res.json(result.rows);
}

// db.execute with ? placeholder
function dbExecuteSafe(req, res) {
    const username = req.body.username;
    db.execute('INSERT INTO logs (username) VALUES (?)', [username]);
    res.sendStatus(200);
}

// Template literal (no interpolation) with ? placeholder
function templateLiteralSafe(req, res) {
    const id = req.query.id;
    connection.query(`SELECT * FROM products WHERE id = ?`, [id], function(err, rows) {
        res.json(rows);
    });
}

File: tests/fixtures/real_world/python/safe/safe_parameterized_variants.py
import sqlite3
import psycopg2
from flask import request

# sqlite3 with ? placeholder
def get_user_sqlite():
    user_id = request.args.get("id")
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM users WHERE id = ?", (user_id,))
    return cursor.fetchone()

# psycopg2 with %s placeholder
def get_user_pg():
    name = request.args.get("name")
    conn = psycopg2.connect("dbname=app")
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM users WHERE name = %s", (name,))
    return cursor.fetchone()

# executemany with ? placeholder
def bulk_insert():
    items = request.json.get("items")
    conn = sqlite3.connect("app.db")
    cursor = conn.cursor()
    cursor.executemany("INSERT INTO items (name) VALUES (?)", items)
    conn.commit()

# conn.execute with ? placeholder
def get_user_conn():
    user_id = request.args.get("id")
    conn = sqlite3.connect("app.db")
    conn.execute("SELECT * FROM users WHERE id = ?", (user_id,))
